File: solutions/haskell/triangle/1/src/Triangle.hs
module Triangle (TriangleType(..), triangleType) where

data TriangleType = Equilateral
                  | Isosceles
                  | Scalene
                  | Illegal
                  deriving (Eq, Show)

triangleType :: (Num a, Ord a) => a -> a -> a -> TriangleType
triangleType a b c 
  | a == b && b == c && isValidTriangle a b c = Equilateral
  | (a == b || b == c || a == c) && isValidTriangle a b c = Isosceles
  | a /= b && b /= c && a /= c && isValidTriangle a b c = Scalene
  | otherwise = Illegal

isValidTriangle :: (Num a, Ord a) => a -> a -> a -> Bool
isValidTriangle a b c = all (> 0) [a,b,c] 
                      && a + b >= c 
                      && b + c >= a 
                      && a + c >= b

File: solutions/python/triangle/2/triangle.py
def equilateral(sides):
    return all(map(lambda x: x == sides[0] and x > 0, sides))


def isosceles(sides):
    return all(map((lambda x: x > 0), sides)) and (sides[0] == sides[1] or sides[0] == sides[2] or sides[1] == sides[2]) and sides[0] + sides[1] >= sides[2] and sides[1] + sides[2] >= sides[0] and sides[0] + sides[2] >= sides[1]


def scalene(sides):
    return all(map((lambda x: x > 0), sides)) and sides[0] != sides[1] and sides[1] != sides[2] and sides[0] != sides[2] and sides[0] + sides[1] >= sides[2] and sides[1] + sides[2] >= sides[0] and sides[0] + sides[2] >= sides[1]
